kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.checksum;
+
+import static io.delta.kernel.internal.actions.SingleAction.CHECKPOINT_SCHEMA;
+import static io.delta.kernel.internal.util.Preconditions.checkState;
+
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.exceptions.ChecksumAlreadyExistsException;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.actions.AddFile;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.replay.CreateCheckpointIterator;
+import io.delta.kernel.internal.stats.FileSizeHistogram;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.stream.Collectors;
+
+public class ChecksumUtils {
+  /**
+   * Computes the state of a Delta table and writes a checksum file based on the provided snapshot.
+   *
+   * <p>This method iterates through the table's data using a CreateCheckpointIterator to calculate:
+   *
+   * <ul>
+   *   <li>Total table size in bytes
+   *   <li>Total number of files
+   *   <li>File size histogram
+   *   <li>Domain metadata information
+   * </ul>
+   *
+   * <p>After computing these statistics, it writes a checksum file to the table's log path.
+   *
+   * <p>Note: For very large tables, this operation may be expensive as in worst case, it requires
+   * scanning all log files until the given version.
+   *
+   * @param engine The Engine instance used to access the underlying storage
+   * @param snapshot The SnapshotImpl instance representing the current state of the table
+   * @throws IOException If an I/O error occurs during checksum computation or writing
+   * @throws ChecksumAlreadyExistsException If a checksum already exists for the snapshot version
+   */
+  public static void computeStateAndWriteChecksum(Engine engine, SnapshotImpl snapshot)
+      throws IOException {
+    if (snapshot.getCurrentCrcInfo().isPresent()) {
+      throw new ChecksumAlreadyExistsException(snapshot.getVersion());
```

Comment:
What's the reason to return an error case here? Can we define this error out of existence? (i.e. this method makes sure that the crc file exists at that version. if it is already there --> great! We can exit early)

https://wiki.tcl-lang.org/page/Define+Errors+Out+of+Existence

Reply:
+1 also removes the need for another exception class

Reply:
Update the method description to say it won't generate if it already exists.

Reply:
Make senses, I just do return early here. IMO if we ignore already exists error here, it is better to also ignore it below